lib/logger.js

```diff
@@ -0,0 +1,653 @@
+'use strict'
+
+const EventEmitter = require('events')
+const os = require('os')
+const querystring = require('querystring')
+const {isIP} = require('net')
+const Agent = require('agentkeepalive')
+const axios = require('axios')
+const stringify = require('json-stringify-safe')
+const constants = require('./constants.js')
+const {checkStringParam, isValidTimestamp, has} = require('./validators/index.js')
+const backoffWithJitter = require('./backoff-with-jitter.js')
+const typeOf = require('./lang/type-of.js')
+
+const kLineLengthTotal = Symbol('lineLengthTotal')
+const kBuffer = Symbol('buffer')
+const kMeta = Symbol('meta')
+const kIsLoggingBackedOff = Symbol.for('isLoggingBackedOff')
+const kFlusher = Symbol('flusher')
+const kAttempts = Symbol('attempts')
+const kRequestDefaults = Symbol.for('requestDefaults')
+const kReadyToSend = Symbol.for('readyToSend')
+const kIsSending = Symbol.for('isSending')
+const kTotalLinesReady = Symbol.for('totalLinesReady')
+const kBackoffMs = Symbol('backoffMs')
+
+const ALL_CLEAR_SENT = 'All accumulated log entries have been sent'
+const ALL_CLEAR_EMPTY = 'All buffers clear; Nothing to send'
+const META_ADD_SUCCESS = 'Successfully added meta property'
+const WARN_LOG_IGNORED = 'Log statement was empty.  Ignored'
+const REMOVE_META_WARN = 'Property is not an existing meta property.  Cannot remove.'
+const REMOVE_META_SUCCESS = 'Successfully removed meta property'
+const ERROR_CODES_TO_RETRY = new Set([
+  500
+, 'ECONNABORTED' // timeout
+])
+
+class Logger extends EventEmitter {
+  constructor(key, opts) {
+    super()
+    const options = opts || {}
+
+    checkStringParam(key, 'LogDNA Ingestion Key', true)
```

Comment:
I think we can be more specific on what we expect to see as the key

Reply:
Specifics like that can be seen in the README (there's a link).  There's no need to do it in code.